### PR TITLE
[Flex] Host failure metrics additions

### LINKED ISSUE #9770: Emit metrics for permanent app failures
We are currently emitting a metric called "AppFailureCount" this metric captures the number of permanent app failures. We are incrementing this count any time a HostInitializationException is thrown in the host. This work item is to expand the scope of where we increment the `AppFailureCount` to capture more app failures. The metrics design doc outlines some scenarios of what we would consider a permanent app failure. Review those scenarios and a) ensure we are incrementing `AppFailureCount` for all the listed use cases and b) discover further permanent app failure use cases and ensure we capture the metrics for these new sceanrios.

### PR DESCRIPTION
**Pull request by satvu**
### Issue describing the changes in this PR

resolves #9770 

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

This covers the original failure scenarios from the design document + the following:

- External startup error (caught by app failure count in Host Initialization from #9730)
- Bundles failure (caught in Host Initialization)
- Unable to parse host configuration file (caught in Host Initialization)

### CODE REVIEW

File: src/WebJobs.Script/Metrics/HostMetricsProvider.cs [hunk 1]
```diff
@@ -96,8 +96,11 @@ private void AddOrUpdateMetricsCache(string key, long value)
             try
             {
                 var functionActivityStatusProvider = _serviceProvider.GetScriptHostServiceOrNull<IFunctionActivityStatusProvider>();
-                var functionActivityStatus = functionActivityStatusProvider.GetStatus();
```
Comment (satvu) [inline src/WebJobs.Script/Metrics/HostMetricsProvider.cs:99]: Bundling this fix in (discussed/fixed with @liliankasem) - this was resulting in a null reference exception in failure scenarios (but was fine with working apps).

File: src/WebJobs.Script.WebHost/Diagnostics/SystemLogger.cs [hunk 2]
```diff
@@ -113,6 +113,7 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
             string source = _categoryName ?? stateSourceValue;
             if (exception is FunctionIndexingException && _eventManager != null)
             {
+                // publish failure metric here?
```
Comment (satvu) [inline src/WebJobs.Script.WebHost/Diagnostics/SystemLogger.cs]: Given this log and [this accumulation of indexing failures](https://github.com/Azure/azure-functions-host/blob/2c830a03dede60feb7aa074dd524d3cb02e7834c/src/WebJobs.Script/Host/ScriptHost.cs#L897), I don't think this is a complete failure/host shutdown, and I'm not sure how we'd want to count this in the metrics. Thoughts?

Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Diagnostics/SystemLogger.cs]: Not a complete host failure. Partial indexing is enabled in Functions, and the host will start with a subset of functions initialized.

**Comment by fabiocav** [comment]
Also had a sync with @satvu offline to cover the other failure scenarios.

**Review by liliankasem — APPROVED** [review]

**Review by fabiocav — APPROVED** [review]

**Pull request merged**